# Week 1 Homework

In [53]:
import pandas as pd
import numpy as np

## Q1: Pandas version
What is the version of the pandas you installed?

In [2]:
pd.__version__

'2.2.3'

We can either download the file from [here](https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv) or simply use `read_csv()` directly.

In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv")

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2160 entries, 0 to 2159
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Laptop        2160 non-null   object 
 1   Status        2160 non-null   object 
 2   Brand         2160 non-null   object 
 3   Model         2160 non-null   object 
 4   CPU           2160 non-null   object 
 5   RAM           2160 non-null   int64  
 6   Storage       2160 non-null   int64  
 7   Storage type  2118 non-null   object 
 8   GPU           789 non-null    object 
 9   Screen        2156 non-null   float64
 10  Touch         2160 non-null   object 
 11  Final Price   2160 non-null   float64
dtypes: float64(2), int64(2), object(8)
memory usage: 202.6+ KB


In [5]:
df.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

## Q2: How many records are in the dataset?
We can simply use `info()` to determine the total rows, or use `len()` or `shape[0]` to give us an actual number of rows.

In [6]:
# determine the number of rows
df.shape[0]

2160

## Q3: Laptop Brands
How many laptop brands are present in the dataset?
- for this, we need to find the unique values under the `Brand` column

In [7]:
len(df["Brand"].unique())

27

There are 27 unique brands available in the dataset

## Q4: Missing Values
How many columns in the dataset have missing values?

In [8]:
for column in df.columns:
    count = df[df[column].isna()].size
    if count > 0:
        print(column)

Storage type
GPU
Screen


There are 3 columns with missing values.

## Q5: Maximum final price
What is the maximum final price of Dell notebooks in the dataset?

In [9]:
max(df[df.Brand=="Dell"]["Final Price"])

3936.0

Maximum final price for Dell notebooks is 3,936

## Q6: Median value of screen
1. Find the median value of Screen column in the dataset.
2. Next, calculate the most frequent value of the same Screen column.
3. Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
4. Now, calculate the median value of Screen once again.

In [32]:
median=float(df.Screen.median())
median

15.6

In [33]:
mode=float(df.Screen.mode().values[0])
mode

15.6

In [34]:
df.Screen.fillna(mode).median()

np.float64(15.6)

There is NO change before and after filling-in missing values, probably since NULL values are excluded from the calculation of the median.

## Q7: Sum of Weights
1. Select all the "Innjoo" laptops from the dataset.
2. Select only columns RAM, Storage, Screen.
3. Get the underlying NumPy array. Let's call it X.
4. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
5. Compute the inverse of XTX.
6. Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
7. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
8. What's the sum of all the elements of the result?

In [40]:
X = df[df.Brand=="Innjoo"][["RAM","Storage","Screen"]].to_numpy()
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [43]:
XTX = np.matmul(X.T,X)
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

In [46]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

In [47]:
y = np.array([1100, 1300, 800, 900, 1000, 1100])
y

array([1100, 1300,  800,  900, 1000, 1100])

In [57]:
v = np.matmul(XTX_inv,X.T)
v

array([[ 0.26298349, -0.12560233, -0.40646389,  0.14958687,  0.05244042,
         0.05244042],
       [-0.00110155,  0.00295059,  0.00125892, -0.00177691, -0.00076387,
        -0.00076387],
       [-0.08772226,  0.0269791 ,  0.17140891, -0.0302108 , -0.00153546,
        -0.00153546]])

In [60]:
w = np.matmul(v,y)
w

array([45.58076606,  0.42783519, 45.29127938])

In [61]:
w.sum()

np.float64(91.2998806299555)

The final sum is 91.30